src/ring_theory/localization.lean

```diff
@@ -1390,6 +1393,97 @@ def localization_localization_at_prime_iso_localization (p : ideal (localization
+  algebra S T :=
+(is_localization.lift (λ y, (map_units T ⟨↑y, h y.prop⟩ : _)) : S →+* T).to_algebra
+
+/-- If `M ≤ N` are submonoids of `R`, then the natural map `M⁻¹S →+* N⁻¹S` commutes with the
+localization maps -/
+lemma localization_is_scalar_tower_of_submonoid_le
+  (M N : submonoid R) (h : M ≤ N) [is_localization M S] [is_localization N T] :
+  @@is_scalar_tower R S T _ (localization_algebra_of_submonoid_le S T M N h).to_has_scalar _ :=
+begin
+  letI := localization_algebra_of_submonoid_le S T M N h,
+  exact is_scalar_tower.of_algebra_map_eq' (is_localization.lift_comp _).symm
+end
+
+noncomputable
+instance (x : ideal R) [H : x.is_prime] [is_domain R] :
+  algebra (localization.at_prime x) (localization (non_zero_divisors R)) :=
+localization_algebra_of_submonoid_le _ _ x.prime_compl (non_zero_divisors R)
+  (by { intros a ha, rw mem_non_zero_divisors_iff_ne_zero, exact λ h, ha (h.symm ▸ x.zero_mem) })
+
+/-- If `M ≤ N` are submonoids of `R`, then `N⁻¹S` is also the localization of `M⁻¹S` at `N`. -/
+lemma is_localization_of_submonoid_le
+  (M N : submonoid R) (h : M ≤ N) [is_localization M S] [is_localization N T] :
+  @@is_localization _ (N.map (algebra_map R S).to_monoid_hom) T _
+    (localization_algebra_of_submonoid_le S T M N h) :=
+begin
+  letI := localization_algebra_of_submonoid_le S T M N h,
+  letI := localization_is_scalar_tower_of_submonoid_le S T M N h,
+  constructor,
+  { rintro ⟨_, ⟨y, hy, rfl⟩⟩,
+    convert is_localization.map_units T ⟨y, hy⟩,
+    exact (is_scalar_tower.algebra_map_apply _ _ _ _).symm },
+  { intro y,
+    obtain ⟨⟨x, s⟩, e⟩ := is_localization.surj N y,
+    refine ⟨⟨algebra_map _ _ x, _, _, s.prop, rfl⟩, _⟩,
+    simpa [← is_scalar_tower.algebra_map_apply] using e },
+  { intros x₁ x₂,
+    obtain ⟨⟨y₁, s₁⟩, e₁⟩ := is_localization.surj M x₁,
+    obtain ⟨⟨y₂, s₂⟩, e₂⟩ := is_localization.surj M x₂,
+    refine iff.trans _ (set.exist_image_iff (algebra_map R S) N (λ c, x₁ * c = x₂ * c)).symm,
+    rw [← (is_localization.map_units T ⟨_, h s₁.prop⟩).mul_left_inj,
+        ← (is_localization.map_units T ⟨_, h s₂.prop⟩).mul_right_inj],
+    simp_rw [is_scalar_tower.algebra_map_apply R S T, ← map_mul],
+    dsimp only [subtype.coe_mk] at e₁ e₂ ⊢,
+    rw [e₁, ← mul_assoc, mul_comm _ x₂, e₂],
+    simp_rw [← map_mul, ← is_scalar_tower.algebra_map_apply R S T],
+    rw is_localization.eq_iff_exists N T,
+    simp only [← (is_localization.map_units S s₁).mul_right_inj] { single_pass := tt },
+    simp only [← @is_unit.mul_right_inj _ _ _ _ (_ * (x₂ * _)) (is_localization.map_units S s₂)]
+      { single_pass := tt },
+    simp only [← mul_assoc] { single_pass := tt },
+    simp only [mul_comm _ x₁, mul_comm _ x₂, ← mul_assoc _ x₂, e₁, e₂, ← map_mul,
+      is_localization.eq_iff_exists M S],
+    split,
+    { rintro ⟨a, e⟩, exact ⟨a, 1, by simpa using e⟩ },
+    { rintro ⟨a, b, e⟩, exact ⟨a * (⟨_, h b.prop⟩ : N), by simpa [mul_assoc] using e⟩ } }
```

Comment:
Here's how I would do the third part of the proof. It's not really shorter, but at least its structure is clearer. Probably the `have h₁, simp only at h₁` part can be extracted into its own lemma (in combination with `is_localization.surj`?)
```suggestion
  constructor,
  { rintro ⟨_, ⟨y, hy, rfl⟩⟩,
    convert is_localization.map_units T ⟨y, hy⟩,
    exact (is_scalar_tower.algebra_map_apply _ _ _ _).symm },
  { intro y,
    obtain ⟨⟨x, s⟩, e⟩ := is_localization.surj N y,
    refine ⟨⟨algebra_map _ _ x, _, _, s.prop, rfl⟩, _⟩,
    simpa [← is_scalar_tower.algebra_map_apply] using e },
  { intros x₁ x₂,
    obtain ⟨⟨y₁, s₁⟩, e₁⟩ := is_localization.surj M x₁,
    obtain ⟨⟨y₂, s₂⟩, e₂⟩ := is_localization.surj M x₂,
    refine iff.trans _ (set.exist_image_iff (algebra_map R S) N (λ c, x₁ * c = x₂ * c)).symm,
    dsimp only at e₁ e₂ ⊢,
    suffices : algebra_map R T (y₁ * s₂) = algebra_map R T (y₂ * s₁) ↔
      ∃ (a : N), algebra_map R S (a * (y₁ * s₂)) = algebra_map R S (a * (y₂ * s₁)),
    { have h₁ := λ x y, @is_unit.mul_left_inj _ _ _ x y (is_localization.map_units T ⟨_, h s₁.prop⟩),
      have h₁' := λ x y, @is_unit.mul_left_inj _ _ _ x y (is_localization.map_units S s₁),
      have h₂ := λ x y, @is_unit.mul_left_inj _ _ _ x y (is_localization.map_units T ⟨_, h s₂.prop⟩),
      have h₂' := λ x y, @is_unit.mul_left_inj _ _ _ x y (is_localization.map_units S s₂),
      simp only [is_scalar_tower.algebra_map_apply R S T, subtype.coe_mk] at h₁ h₂,
      simp only [is_scalar_tower.algebra_map_apply R S T, map_mul, ← e₁, ← e₂, ← mul_assoc,
        mul_right_comm _ (algebra_map R S s₂),
        mul_right_comm _ (algebra_map S T (algebra_map R S s₂)),
        h₂, h₁, h₂', h₁'] at this,
      simpa only [mul_comm] using this },
    simp_rw [is_localization.eq_iff_exists N T, is_localization.eq_iff_exists M S],
    split,
    { rintro ⟨a, e⟩, exact ⟨a, 1, by { convert e using 1; simp; ring }⟩ },
    { rintro ⟨a, b, e⟩, exact ⟨a * (⟨_, h b.prop⟩ : N), by { convert e using 1; simp; ring }⟩ } }
```